ruby_on_rails/testing/model_testing.md

```diff
@@ -0,0 +1,132 @@
+
+## Creating a factory
+
+Below is an example of a user factory. Notice that the factory will be a User class based off of the name of your factory. You can also define as many attributes as you want and it can be useful to focus on adding only the attributes your testing and adding them as need be.
+
+~~~
+
+# This will guess the User class
+FactoryBot.define do
+  factory :user do
+    first_name { "John" }
+    last_name  { "Doe" }
+    admin { false }
+  end
+end
+
+~~~
+
+When creating a factory bot record you have the choice of using the `create(:user)` which returns a saved instance of the object from your database or `build(:user)` which returns an unsaved instance of an object. You might be wondering how we avoid objects hanging around in the database. Thankfully RSpec includes the ability to clear the database in between tests. However if you do find that RSpec's database clearing isn't fitting your needs here is a alternative.
+
+navigating to the Gemfile we will want to install the `database_cleaner` gem and this is what we will use to clean our databases between our test runs. Navigate to the Gemfile and add the following gem to the test and development group then in your console enter `bundle install` to add the Gem.
+
+~~~
+
+group :development, :test do
+  gem 'database_cleaner-active_record' #Add this line. 
+end
+
+~~~
+
+You will also need to navigate to your `rails_helper` file and change `config.use_transactional_fixtures` to false. Then go to the `support` folder in your spec folder and create a `database_cleaner.rb` file. This is where we will add the following configuration to clear out the test data.
+
+~~~
+
+RSpec.configure do |config|
+    config.before(:suite) do
+      DatabaseCleaner.clean_with(:truncation)
+      Rails.application.load_seed
+    end
+  
+    config.before(:each) do
+      DatabaseCleaner.strategy = :transaction
+    end
+  
+    config.before(:each, :js => true) do
+      DatabaseCleaner.strategy = :truncation
+    end
+  
+    config.before(:each) do
+      DatabaseCleaner.start
+    end
+  
+    config.after(:each) do
+      DatabaseCleaner.clean
+    end
+  end
+
+~~~
+
+  Now your finally done configuring your tests and can set up your first one!
+
+## Example model test
+
+When testing our models we usually want to focus on testing the business logic. Essentially what goes on behind each model that enables the user experience to work correctly. This could mean ensuring promotion tickets prices are correct, only correct form data is accepted, custom user verification etc. In the context of a user it would like like the following. It's important to note that each model should have it's own file in the `spec/models` file. This is helpful for keeping your model tests organized and readable.
+
+~~~
+
+RSpec.describe User, type: :model do
+  it "verifies the user has a number."
+    user = build(:user, user_name: "Forbidden username!") 
+    expect(user.save). to eql(false)
+  end
+end
+
+~~~
```

Comment:
I am not sure if this test is very clear. Maybe the explanation needs to come first, but I can't tell if we are testing if the name is valid, or if the user has a number.

I also think it could be useful here to build out 2-3 tests here, using contexts to show different examples of it being true and false.